app/Models/Beatmapset.php

```diff
@@ -661,6 +679,74 @@ public function beatmaps()
         return $this->hasMany(Beatmap::class, 'beatmapset_id');
     }
 
+    public function events()
+    {
+        return $this->hasMany(BeatmapsetEvent::class);
+    }
+
+    public function requiredNominationCount()
+    {
+        $longest_map_duration = $this->beatmaps->max('total_length');
+
+        return $longest_map_duration > 315 ? 3 : 2;
+    }
+
+    public function currentNominationCount()
+    {
+        return count($this->recentEvents()->nominations()->get());
+    }
+
+    public function rankingETA()
+    {
+        if (!$this->isQualified()) {
+            return;
+        }
+
+        $queueSize = static::qualified()->where('approved_date', '<', $this->approved_date)->count();
+        $days = ceil($queueSize / static::QUALIFICATIONS_PER_DAY);
+
+        return $days > 0 ? Carbon::now()->addDays($days)->startOfDay() : null;
+    }
+
+    public function recentEvents()
+    {
+        // relevant events differ depending on state of beatmapset
+        $events = $this->events();
+        switch ($this->approved) {
+            case self::STATES['pending']:
+            case self::STATES['qualified']:
+                // last 'disqualify' event (if any) and all events since
+                $disqualifyEvent = $this->events()->disqualifications()->orderBy('created_at', 'desc')->first();
+                if ($disqualifyEvent) {
+                    $events = $events->where('id', '>=', $disqualifyEvent->id);
```

Comment:
(which is why I said you shouldn't use `$event` for `$disqualifyEvent` above - `$disqualifyEvent = $event->disqualifications()` changes the `$event` as well)